src/Ersatz/Relation/Set.hs

```diff
@@ -0,0 +1,435 @@
+
+-- | Construct a set \(S \subseteq A \) from a list.
+build :: (Ix a)
+  => (a,a)
+  -> [(a,Bit)]
+  -> Set a
+build =
+  Set
+  . uncurry RD.build
+  . (relBounds_ *** fmap relAssoc_)
+  .: (,)
+
+-- | Construct a set \(S \subseteq A \) from a predicate.
+buildFrom :: (Ix a)
+  => (a,a)
+  -> (a -> Bit)
+  -> Set a
+buildFrom =
+  Set
+  . uncurry RD.buildFrom
+  . (relBounds_ *** lift_)
+  .: (,)
+
+-- | Construct an indeterminate set \(S \subseteq A \) from a predicate.
+buildFromM :: (Ix a, MonadSAT s m)
+  => (a,a)
+  -> (a -> m Bit)
+  -> m (Set a)
+buildFromM =
+  fmap Set
+  . uncurry RD.buildFromM
+  . (relBounds_ *** lift_)
+  .: (,)
+
+-- | Construct a symbolic set \( \Omega \subseteq A \) from bounds where, by
+-- construction, every element within the bounds is contained in the set.
+universe' :: (Ix a) => (a,a) -> Set a
+universe' =
+  build
+  <*> map (, true) . Ix.range
+
+-- | Construct a symbolic set \( \varnothing \subseteq A \) from bounds where, by
+-- construction, no elements are contained in the set.
+empty :: (Ix a) => (a,a) -> Set a
+empty =
+  build
+  <*> map (, false) . Ix.range
+
+-- | Construct a symbolic set \( \{a\} \subseteq A \) from bounds where, by
+-- construction, exactly one particular element \( a \) is contained in the set.
+singleton :: (Ix a) => (a,a) -> a -> Set a
+singleton bnd a =
+  build bnd
+  $ map (id &&& (bool . (a ==)))
+        $ Ix.range bnd
+
+
+
+-- Concrete queries
+
+-- | The bounds of the array representing the set.
+bounds :: (Ix a) => Set a -> (a,a)
+bounds = unRelBounds_ . RD.bounds . coerce
+
+-- | The list of indices of the array representing the set.
+indices :: (Ix a) => Set a -> [a]
+indices = fmap unPad . RD.indices . coerce
+
+-- | The list of tuples for the given set \(S \subseteq A \), where the first
+-- element is an index and the second element is a 'Bit' indicating whether the
+-- element is contained in the represented set.
+assocs :: (Ix a) => Set a -> [(a, Bit)]
+assocs = fmap unRelAssoc_ . RD.assocs . coerce
+
+-- | The list of elements of the array that represents the given set.
+elems :: (Ix a) => Set a -> [Bit]
+elems = RD.elems . coerce
+
+-- | Synonym for 'indices', provided for parallelism with 'Data.Relation'.
+universe :: (Ix a) => Set a -> [a]
+universe = RD.domain . coerce
+
+-- | The size of the universe that a symbolic (sub)set is a subset of.
+universeSize :: (Ix a) => Set a -> Int
+universeSize = A.rangeSize . bounds
+
+
+
+-- Symbolic queries/assertions
+
+-- | The number of elements \( |S| \) contained in the given symbolic subset
+-- \( S \subseteq A \).
+card :: (Ix a) => Set a -> Bits
+card = RD.card . coerce
+
+-- | @s ! a@ tests if \( a \in S \).
+(!) :: (Ix a) => Set a -> a -> Bit
+(!) =
+  uncurry (RD.!)
+  . (coerce *** pad)
+  .: (,)
+
+-- | Flipped, non-infix synonym for '!'.
+member :: (Ix a) => a -> Set a -> Bit
+member = flip (!)
+
+-- | Tests if the set is empty.
+null :: (Ix a) => Set a -> Bit
+null = RP.empty . coerce
```

Comment:
Mismatch with the existing `Relation` interface: `Relation.Prop` currently uses `empty` for the equivalent predicate. 

It is much more idiomatic for `null` to correspond to an emptiness predicate and `empty` to correspond to an empty collection constructor, and that's what I've gone with here. I'd suggest changing the existing `Relation` API to match as well.

Reply:
I'm not sure what the history of naming conventions in Haskell is, but I was thinking more about `Foldable` than `containers` per se.

Nevertheless, I do think consistency with other Haskell conventions is generally desirable, however, and am happy to change `add`/`del`to match `insert`/`delete`.

A more significant changes to the `Set` — and `Relation` API — to bring things in line with other conventions would be to distinguish predicates from operations by prefixing predicates with `is/are_`. More important than adherence to convention per se, this would permit both a predicate (often easier to correctly define) and a related operation to live in the same namespace unqualified.

Reply:
I've updated the `Set` API to align significantly better with what someone would expect given `Data.Set` and updated `Ersatz.Relation` to match.

Reply:
@RyanGlScott any chance you've had some time to offer feedback on the updates here?